src/deroff.rs

```diff
@@ -538,6 +538,307 @@ impl Deroffer {
+                self.s = self.skip_char(self.s.as_str(), None).to_owned();
+                // return self.text();
+            },
+            "." => {
+                self.r#macro = false;
+                // self.condputs("\n");
+                return true;
+            },
+            _ => {()}
+        };
+
+        self.nobody = false;
+        let s0s1 = self.s[..2].to_owned();
+
+        if self.g_macro_dispatch(s0s1.as_str()) {
+            return true;
+        }
+
+        if SKIP_HEADERS && self.nobody {
+            return true;
+        }
+
+        self.s = self.skip_leading_whitespace(self.s.as_str()).to_owned();
+        while !self.s.is_empty() && !Self::is_white(self.s.as_str(), 0) {
+            self.s = self.skip_char(self.s.as_str(), None).to_owned();
+        }
+
+        self.s = self.skip_leading_whitespace(self.s.as_str()).to_owned();
+
+        loop {
+            if !self.quoted_arg() && !self.text_arg() {
+                if !self.s.is_empty() {
+                    // self.condputs(self.str_at(0));
+                    self.condputs(Self::str_at(self.s.as_str(), 0));
+                    self.s = self.skip_char(self.s.as_str(), None).to_owned();
+                } else {
+                    return true;
+                }
+            }
+        }
+    }
+
+    fn font(&mut self) -> bool {
+        if let Some(m) = self.g_re_font.find(self.s.as_str()) {
+            self.s = self.skip_char(self.s.as_str(), Some(m.end())).to_owned();
+            true
+        } else {
+            false
+        }
+    }
+
+    fn comment(&self) -> bool {
+        unimplemented!()
+    }
+
+    fn numreq(&mut self) -> bool {
+        // In the python, it has a check that is already handled in esc_char_backslash, which is 
+        // the only place it gets called, so I'll omit that check here
+
+        // This is written as `self.macro += 1` in the source, but I dont know why
+        // it does the same thing (false -> true, true -> still true) :shrug:
+        // self.r#macro = true;
+        // Upon further investigation, this is the weirdest function ever
+        // This is just a state placeholder thing
+        let mut m = self.r#macro as u8 + 1;
+
+        // self.skip_char(3);
+        self.s = self.skip_char(self.s.as_str(), Some(3)).to_owned();
+
+        // There's this, which has a comment explaining my thoughts right now
+        /* 
+        while self.str_at(0) != "'" and self.esc_char():
+            pass  # Weird
+        */
+        // And I dont know what the purpose of it would even be, if you can tell, lmk
+
+        // if self.str_at(0) == '\'' {
+        if Self::str_at(self.s.as_str(), 0) == "'" {
+            // self.skip_char(1);
+            self.s = self.skip_char(self.s.as_str(), None).to_owned();
+        }
+
+        m -= 1;
+        self.r#macro = m != 0;
+
+        true
+    }
+
+    fn text_arg(&self) -> bool {
+        unimplemented!()
+    }
+
+    fn var(&mut self) -> bool {
```

Comment:
Yikes! This is a gross bear of a function! Nice work!

Need a test function with a decent number of assertions to show what it's doing and prevent us from breaking it in the future.

Reply:
This is the worst function in this entire file. I'm glad you said to go back and add tests. There's a solid amount of comments in the tests explaining what the code is doing, as well as a python version of the tests to run with the python source to make sure it perfectly mirrors the behavior. It's not perfect yet, `text_arg` needs to be implemented, but after that, we never have to look here ever again.
c40304f